packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -143,14 +143,39 @@ export default createRule<Options, MessageId>({
 
     function nodeIsArrayType(node: TSESTree.Expression): boolean {
       const nodeType = getNodeType(node);
-      return checker.isArrayType(nodeType) || checker.isTupleType(nodeType);
+      return checker.isArrayType(nodeType);
+    }
+    function nodeIsTupleType(node: TSESTree.Expression): boolean {
+      const nodeType = getNodeType(node);
+      return checker.isTupleType(nodeType);
+    }
+
+    function isArrayIndexExpression(node: TSESTree.Expression): boolean {
+      return (
+        // Is an index signature
+        node.type === AST_NODE_TYPES.MemberExpression &&
+        node.computed &&
+        // ...into an array type
+        (nodeIsArrayType(node.object) ||
+          // ... or a tuple type
+          (nodeIsTupleType(node.object) &&
+            // Exception: literal index into a tuple - will have a sound type
+            node.property.type !== AST_NODE_TYPES.Literal))
+      );
     }
```

Comment:
won't this also match non-element access?
i.e.
```
arr['length']
```

Thinking about it - it doesn't actually matter, because there are no nullish properties on an array.